core/src/main/scala/akka/kafka/scaladsl/Consumer.scala

```diff
@@ -99,6 +99,15 @@ object Consumer {
     Source.fromGraph(ConsumerStage.plainSubSource[K, V](settings, subscription))
 
   /**
+   * The `plainPartitionedManualOffsetSource` is similar to [[#plainPartitionedSource]] but allows the use of an offset store outside
+   * of Kafka, while retaining the automatic partition assignment. When a topic-partition is assigned to a consumer, the `loadOffsetsOnAssign`
+   * function will be called to retrieve the offset, followed by a seek to the correct spot in the partition. The `onRevoke` function gives
+   * the consumer a chance to store any uncommitted offsets, and do any other cleanup that is required.
+   */
+  def plainPartitionedManualOffsetSource[K, V](settings: ConsumerSettings[K, V], subscription: AutoSubscription, loadOffsetOnAssign: Set[TopicPartition] => Map[TopicPartition, Long], onRevoke: Set[TopicPartition] => Unit = _ => ()): Source[(TopicPartition, Source[ConsumerRecord[K, V], NotUsed]), Control] =
```

Comment:
For the record, we discussed the possibility to let `loadOffsetOnAssign` return a `Future[Map[TopicPartition, Long]]` instead to better support loading the offset asynchronously from an external place. The reason that doesn't work is because `seek` must be called before the `onPartitionsAssigned` callback is returned. Also note that `onPartitionsAssigned` is called on the same thread as called `poll`, i.e. the KafkaConsumerActor.

Reply:
btw, it should be renamed to `getOffsetOnAssign`

Reply:
I've spent some more time digging into the `KafkaConsumer` code and I think I have a better understanding of what's happening now. I think we will actually be able to go back to accepting a `Future[Map[TopicPartition, Long]]` from `getOffsetOnAssign`. The `WrappedAutoPausedListener` pauses a partition as soon as it is assigned, so as long as that is done synchronously, we're free to do the `seek` and then `pump` asynchronously as long as we do them in that order, since a partition will only be `resume`d when a request comes to the `KafkaConsumerActor` from a downstream `Source`, which is created in `pump()`.